# Business Understanding and Set-up

## Background and Key Question

**Airbnb**

Airbnb is an American online platform offering vacation rentals for travelers - primarily homestays - as an alternative to traditional hotel or hostel stays. Additionally, as of late Airbnb also offers "experiences" at popular tourist destinations. For both listings and experiences, Airbnb does not own or host themselves but instead act as an intermediary broker, earning commissions for each booking. Consequently, as many other groundbreaking innovative services Airbnb has not been without controversies, as it offers owners and renters with a great opportunity to make use of unused periods, but incentivizes the dedicated (ab)use of living space as full-time rentals.

**Key Question**

Taking data from a given specific listing time stamp of Berlin, **can we accurately predict its price** in order to provide future hosts with a solid pricing estimate without requiring an Airbnb account beforehand? (aim: offer a tool for new users to get a first pricing and potential earnings indication without requiring an account)

**Assumptions**

As the data is accessible information only and does not include data such as actual occupancy, several assumptions were necessary and prediction could not possibly be conducted with a precision matching a state of complete information. Nevertheless, the wealth of available features made it possible to experiment extensively and work productively on the basis of these assumptions.

**Note**: The final key question was derived as an iterative process, adapting the target based on the insights from the analysis. Hence, the analysis focuses on evaluating all feasible features, as the decision to include only those that are known to a new user beforehand was made further down the road.

## Feature Glossary

This glossary only lists the key_features used for the final model prediction and does not cover all 105 original features of the dataset.

| **FEATURE** | **DESCRIPTION** |
| :----- | :----- |
| **TARGET: price_log** | log of price per guests_included for a listing |
| **accommodates_per_bed** | created from "accommodates" and "beds" - how many people the property accommodates per available bed |
| **am_balcony** | created from "amenities" - does the listing have a balcony? |
| **am_breakfast** | created from "amenities" - does the rent include breakfast? |
| **am_child_friendly** | created from "amenities" - is the listing flagged as child friendly? |
| **am_elevator** | created from "amenities" - does the listing have an elevator? |
| **am_essentials** | created from "amenities" - does the host provide essentials (e.g. towels, bed linen, ...)? |
| **am_pets_allowed** | created from "amenities" - are pets allowed at the location? |
| **am_private_entrance** | created from "amenities" - do guests have a private entrance? |
| **am_smoking_allowed** | created from "amenities" - is smoking allowed at the location? |
| **am_tv** | created from "amenities" - is a TV available? |
| **bathrooms_log** | log of number of shared (0.5) or dedicated (1.0) bathrooms available |
| **bedrooms** | number of shared (0.5) or dedicated (1.0) bedrooms available |
| **calc_host_lst_count_sqrt_log** | Log of sqrt of number of listings the host owns in total |
| **cancellation_policy** | "flexible", "moderate", "strict" or "super strict" cancellation policies set by the host |
| **guests_included_calc** | number of guests included in the listing price (not necessarily equal to "accommodates", as host may charge for extra_people) |
| **host_is_superhost** | does the host fulfill the criteria and is flagged as a superhost (e.g. high ratings, response rate, ...)? |
| **instant_bookable** | is the listing instant bookable (e.g. without manual confirmation by the host)? |
| **maximum_nights** | what are the maximum  nights specified by the host? |
| **minimum_nights_sqrt** | log of what are the minimum nights specified by the host? |
| **property_type** | which type of property is the listing (e.g. apartment, house, hotel, ...)? |
| **room_type** | which type of room is listed (e.g. entire home/apt, private room, shared room, ...)? |
| **wk_mth_discount** | created from weekly_price and monthly_price - does the host offer a discount for weekly or monthly stays (in %)? |
| **zipcode** | at which zipcode is the listing located? |

## Dataset Glossary

| **DATASET** | **DESCRIPTION** |
| :----- | :----- |
| **data_raw** | Originally imported dataset listings.csv.gz (February 2020) |
| **data** | Naming for main working dataset throughout all notebooks |
| **data_clean** | Export from Notebook 1_Clean, import for Notebooks 2_EDA_Clean and 3_Feature_Engineering |
| **data_engineered** | Export from Notebook 3_Feature_Engineering, import for Notebook 4-EDA_Engineered and 5_Predictive_Modeling |
| **best_model_xyz** | Saves of best models from various algorithms |



## Target Feature(s) and Metric(s)

**Target**:
- Feature: price_log
- Metric: neg_median_absolute_error

## Libraries and Dashboard

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from numpy import loadtxt
import os, glob
from datetime import datetime
from datetime import timedelta
%matplotlib inline

In [2]:
# Dashboard
dataset_loc = "berlin"  # "berlin", "paris", "amsterdam"
dataset_date = "2020-03-17"  # "2019-12-11", "2020-01-10", "2020-02-18", "2020-03-17", "2020-05-14"
dataset_date_min3mth = "2019-12-11"
pd.set_option('display.max_columns', 150)
pd.set_option('display.max_rows', 100)
pd.options.display.max_seq_items = 300
sns.set(style="white")

# Data Mining

## Data Checks

The monthly data is composed of various files that are briefly visualized here (based on dataset_loc and dataset_date):

- listings.csv.gz
- listings.csv
- reviews.csv.gz
- reviews.csv
- calendar.csv.gz
- neighbourhoods.csv
- neighbourhoods.geojson

**listings.csv.gz**

In [3]:
# Display contents of listings.csv.gz as well as its shape
#data_listings_gz_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/listings.csv.gz")
#print(data_listings_gz_insp.shape)
#data_listings_gz_insp.head(3)

**listings.csv**

In [4]:
# Display contents of listings.csv as well as its shape
#data_listings_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/listings.csv")
#print(data_listings_insp.shape)
#data_listings_insp.head(2)

**reviews.csv.gz**

In [5]:
# Display contents of reviews.csv.gz as well as its shape
#data_reviews_gz_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/reviews.csv.gz")
#print(data_reviews_gz_insp.shape)
#data_reviews_gz_insp.head(2)

**reviews.csv**

In [6]:
# Display contents of reviews.csv as well as its shape
#data_reviews_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/reviews.csv")
#print(data_reviews_insp.shape)
#data_reviews_insp.head(2)

**calendar.csv.gz**

In [7]:
# Display contents of calendar.csv.gz as well as its shape
#data_cal_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/calendar.csv.gz")
#print(data_cal_insp.shape)
#data_cal_insp.head(2)

**neighbourhoods.csv**

In [8]:
# Display contents of neighbourhoods.csv as well as its shape
#data_neighb_insp = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/neighbourhoods.csv")
#print(data_neighb_insp.shape)
#data_neighb_insp.head(2)

**neighbourhoods.geojson**

In [9]:
# Display contents of neighbourhoods.geojson as well as its shape
#data_neighb_geojson_insp = gpd.read_file(f"data/{dataset_loc}_{dataset_date}/neighbourhoods.geojson")
#print(data_neighb_geojson_insp.shape)
#data_neighb_geojson_insp.head(2)

## Data Import

**Create main dataset (listings on January 10th, i.e. pre-COVID-19)**

In [10]:
# Import dataset as DataFrame (as csv-file)
data_raw = pd.read_csv(f"data/{dataset_loc}_{dataset_date}/listings.csv.gz")

C:\Users\Mauricio\anaconda3\envs\airbnb42\lib\site-packages\IPython\core\interactiveshell.py:3072: DtypeWarning: Columns (43,61,62) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [11]:
# Assign data_raw to data (in order to always keep a freshly imported data_raw) and set id as index
data = data_raw.copy()
data.set_index('id', inplace=True)

# Data Cleaning

## Pre-cleaning

Before cleaning the main dataset, the columns were briefly reviewed and all those that were deemed unfruitful for further analysis eliminated beforehand, in order to focus time and effort on the relevant ones.

In [12]:
# Display shape of "data"
data.shape

(25164, 105)

The dataset includes a whopping **105 features** and slightly over 25.000 listings for Berlin for March 2020.

In [13]:
# Display head(1) of "data"
data.head(1)

listing_url       scrape_id last_scraped  \
id                                                                     
3176  https://www.airbnb.com/rooms/3176  20200317045838   2020-03-17   

                                 name  \
id                                      
3176  Fabulous Flat in great Location   

                                                summary  \
id                                                        
3176  This beautiful first floor apartment  is situa...   

                                                  space  \
id                                                        
3176  1st floor (68m2) apartment on Kollwitzplatz/ P...   

                                            description experiences_offered  \
id                                                                            
3176  This beautiful first floor apartment  is situa...                none   

                                  neighborhood_overview  \
id                                                        
3176  The neighbourhood is famous for its variety of...   

                                                  notes  \
id                                                        
3176  We welcome FAMILIES and cater especially for y...   

                                                transit  \
id                                                        
3176  We are 5 min walk  away from the tram M2, whic...   

                                                 access  \
id                                                        
3176  The apartment will be entirely yours. We are c...   

                                            interaction  \
id                                                        
3176  Feel free to ask any questions prior to bookin...   

                                            house_rules  thumbnail_url  \
id                                                                       
3176  It’s a non smoking flat, which likes to be tre...            NaN   

      medium_url                                        picture_url  \
id                                                                    
3176         NaN  https://a0.muscache.com/im/pictures/243355/84a...   

      xl_picture_url  host_id                                host_url  \
id                                                                      
3176             NaN     3718  https://www.airbnb.com/users/show/3718   

     host_name  host_since                         host_location  \
id                                                                 
3176    Britta  2008-10-19  Coledale, New South Wales, Australia   

                                             host_about  host_response_time  \
id                                                                            
3176  We love to travel ourselves a lot and prefer t...  within a few hours   

     host_response_rate host_acceptance_rate host_is_superhost  \
id                                                               
3176               100%                  80%                 f   

                                     host_thumbnail_url  \
id                                                        
3176  https://a0.muscache.com/im/users/3718/profile_...   

                                       host_picture_url host_neighbourhood  \
id                                                                           
3176  https://a0.muscache.com/im/users/3718/profile_...    Prenzlauer Berg   

      host_listings_count  host_total_listings_count  \
id                                                     
3176                  1.0                        1.0   

                                     host_verifications host_has_profile_pic  \
id                                                                             
3176  ['email', 'phone', 'facebook', 'reviews', 'man...                    t   

     host_identity_verified                   street    neighbourhood  \
id      

On first sight, most features appear relatively informative and well-structured. **Amenities** stores lists as a string, which needs to be reviewed later on.

In [14]:
# Display columns of "data"
#data.columns

In [15]:
# Define columns to keep after pre-cleaning
select_columns = [
    'accommodates', 'amenities', 'availability_365', 'availability_90',
    'bathrooms', 'bed_type', 'bedrooms', 'beds',
    'calculated_host_listings_count', 'cancellation_policy', 'cleaning_fee',
    'description', 'experiences_offered', 'extra_people', 'first_review',
    'guests_included', 'has_availability', 'host_acceptance_rate',
    'host_has_profile_pic', 'host_identity_verified', 'host_is_superhost',
    'host_listings_count', 'host_location', 'host_response_rate',
    'host_response_time', 'house_rules', 'instant_bookable', 'interaction',
    'is_business_travel_ready', 'is_location_exact', 'last_review', 'latitude',
    'listing_url', 'longitude', 'maximum_nights', 'minimum_nights',
    'monthly_price', 'name', 'neighborhood_overview', 'neighbourhood_cleansed',
    'notes', 'number_of_reviews', 'number_of_reviews_ltm', 'price',
    'property_type', 'require_guest_phone_verification',
    'require_guest_profile_picture', 'requires_license',
    'review_scores_accuracy', 'review_scores_checkin',
    'review_scores_cleanliness', 'review_scores_communication',
    'review_scores_location', 'review_scores_rating', 'review_scores_value',
    'reviews_per_month', 'room_type', 'security_deposit', 'space',
    'square_feet', 'summary', 'transit', 'weekly_price', 'zipcode'
]

In [16]:
# Drop innecessary columns and sort dataset
drop_columns = [el for el in data.columns if el not in select_columns]
data.drop(labels=drop_columns, inplace=True, axis=1)
data = data.reindex(sorted(data.columns, reverse=False), axis=1)

As mentioned above, around half of the features are kept for further analysis and the rest is dropped - at least for the time being

## Inspection

In [17]:
# Display shape of "data"
data.shape

(25164, 64)

The feature set is now reduced to **64 features**

In [18]:
# Display head(5) of remaining "data"
data.head(5)

accommodates                                          amenities  \
id                                                                      
3176             4  {Internet,Wifi,Kitchen,"Buzzer/wireless interc...   
3309             1  {Internet,Wifi,"Pets live on this property",Ca...   
6883             2  {TV,"Cable TV",Internet,Wifi,"Air conditioning...   
7071             2  {Wifi,Heating,"Family/kid friendly",Essentials...   
9991             7  {TV,"Cable TV",Internet,Wifi,Kitchen,"Paid par...   

      availability_365  availability_90  bathrooms       bed_type  bedrooms  \
id                                                                            
3176               140                0        1.0       Real Bed       1.0   
3309               320               45        1.0  Pull-out Sofa       1.0   
6883                 0                0        1.0       Real Bed       1.0   
7071                45               45        1.0       Real Bed       1.0   
9991                 8                0        2.5       Real Bed       4.0   

      beds  calculated_host_listings_count          cancellation_policy  \
id                                                                        
3176   2.0                               1  strict_14_with_grace_period   
3309   1.0                               1  strict_14_with_grace_period   
6883   1.0                               1                     moderate   
7071   2.0                               2                     moderate   
9991   7.0                               1  strict_14_with_grace_period   

     cleaning_fee                                        description  \
id                                                                     
3176      $100.00  This beautiful first floor apartment  is situa...   
3309       $30.00  First of all: I prefer short-notice bookings. ...   
6883       $39.00  Stay in a stylish loft on the second floor and...   
7071        $0.00  Cozy and large room in the beautiful district ...   
9991       $80.00  4 bedroom with very large windows and outstand...   

     experiences_offered extra_people first_review  guests_included  \
id                                                                    
3176                none       $20.00   2009-06-20                2   
3309                none       $18.00   2013-08-12                1   
6883                none        $0.00   2010-02-15                1   
7071                none       $27.00   2009-08-18                1   
9991                none       $10.00   2015-08-09                5   

     has_availability host_acceptance_rate host_has_profile_pic  \
id                                                                
3176                t                  80%                    t   
3309                t                 100%                    t   
6883                t                 100%                    t   
7071                t                  96%                    t   
9991                t                  25%                    t   

     host_identity_verified host_is_superhost  host_listings_count  \
id                                                                   
3176                      t                 f                  1.0   
3309                      f                 f                  1.0   
6883                      t                 f                  1.0   
7071                      t                 t                  2.0   
9991                      t                 f                  1.0   

                             host_location host_response_rate  \
id                                                              
3176  Coledale, New South Wales, Australia               100%   
3309               Berlin, Berlin, Germany               100%   
6883               Berlin, Berlin, Germany               100%   
7071               Berlin, Berlin, Germany               100%   
9991               Berlin, Berlin, Germany               100

In [19]:
# Describe data (summary)
data.describe().round(2).T

count    mean     std    min    25%  \
accommodates                    25164.0    2.70    1.60   1.00   2.00   
availability_365                25164.0   72.81  112.20   0.00   0.00   
availability_90                 25164.0   21.90   31.32   0.00   0.00   
bathrooms                       25146.0    1.10    0.35   0.00   1.00   
bedrooms                        25131.0    1.16    0.68   0.00   1.00   
beds                            24951.0    1.61    1.23   0.00   1.00   
calculated_host_listings_count  25164.0    2.44    5.44   1.00   1.00   
guests_included                 25164.0    1.37    0.92   1.00   1.00   
host_listings_count             25142.0    3.92   39.34   0.00   1.00   
latitude                        25164.0   52.51    0.03  52.34  52.49   
longitude                       25164.0   13.41    0.06  13.10  13.37   
maximum_nights                  25164.0  597.86  548.36   1.00  20.00   
minimum_nights                  25164.0    7.14   26.74   1.00   2.00   
number_of_reviews               25164.0   21.69   46.84   0.00   1.00   
number_of_reviews_ltm           25164.0    7.55   15.87   0.00   0.00   
review_scores_accuracy          20110.0    9.68    0.75   2.00  10.00   
review_scores_checkin           20089.0    9.73    0.70   2.00  10.00   
review_scores_cleanliness       20113.0    9.33    1.04   2.00   9.00   
review_scores_communication     20108.0    9.75    0.69   2.00  10.00   
review_scores_location          20090.0    9.60    0.73   2.00   9.00   
review_scores_rating            20137.0   94.63    7.76  20.00  93.00   
review_scores_value             20087.0    9.41    0.86   2.00   9.00   
reviews_per_month               20636.0    1.02    1.48   0.01   0.14   
square_feet                       400.0  497.23  521.19   0.00   0.00   

                                    50%      75%       max  
accommodates                       2.00     3.00     24.00  
availability_365                   0.00   101.00    365.00  
availability_90                    0.00    45.00     90.00  
bathrooms                          1.00     1.00      8.50  
bedrooms                           1.00     1.00     12.00  
beds                               1.00     2.00     24.00  
calculated_host_listings_count     1.00     2.00     57.00  
guests_included                    1.00     1.00     24.00  
host_listings_count                1.00     2.00   1384.00  
latitude                          52.51    52.53     52.66  
longitude                         13.42    13.44     13.76  
maximum_nights                  1120.00  1125.00  10000.00  
minimum_nights                     3.00     4.00   1124.00  
number_of_reviews                  5.00    18.00    940.00  
number_of_reviews_ltm              1.00     7.00    590.00  
review_scores_accuracy            10.00    10.00     10.00  
review_scores_checkin             10.00    10.00     10.00  
review_scores_cleanliness         10.00    10.00     10.00  
review_scores_communication       10.00    10.00     10.00  
review_scores_location            10.00    10.00     10.00  
review_scores_rating              97.00   100.00    100.00  
review_scores_value               10.00    10.00     10.00  
reviews_per_month                  0.44     1.26     34.33  
square_feet                      431.00   753.00   4639.00

Data.describe gives a positive impression of the dataset quality. Certain features have missing values and it is noticeable that not all features which should have numeric values show up, which indicates that there may be some issues with datatypes.

In [20]:
# List datatypes (data.info()) (pre-cleaning)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25164 entries, 3176 to 42927052
Data columns (total 64 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   accommodates                      25164 non-null  int64  
 1   amenities                         25164 non-null  object 
 2   availability_365                  25164 non-null  int64  
 3   availability_90                   25164 non-null  int64  
 4   bathrooms                         25146 non-null  float64
 5   bed_type                          25164 non-null  object 
 6   bedrooms                          25131 non-null  float64
 7   beds                              24951 non-null  float64
 8   calculated_host_listings_count    25164 non-null  int64  
 9   cancellation_policy               25164 non-null  object 
 10  cleaning_fee                      17725 non-null  object 
 11  description                       24664 non-null  object 
 12

As expected, various numerical features are currently stored as objects and need to be transformed (e.g. cleaning_fee, extra_people, first_review, price, ...). Additionally, there are actually quite a few columns with missing values.

In [21]:
# Show maximum/minimum value for each numerical column
#num_features = list(data.columns[data.dtypes!=object])
#data[num_features].max()
#data[num_features].min()

Several rows with unusually high values can be identified and may in some cases be dropped at a certain threshold during data handling. Some particular features include:

| **FEATURE** | **MAX_VALUE** |
| :----- | :----- |
| **calculated_host_listings_count** | 57 |
| **accommodates** | 24 |
| **bedrooms** | 12 |
| **beds** | 24 |
| **minimum_nights** | 1.124 |
| **maximum_nights** | 10.000 |
| **number_of_reviews_ltm** | 590 (potentially misleading; actually had less reviews on Airbnb |

In [22]:
# List unique entries per column
data.nunique()

accommodates                           17
amenities                           22601
availability_365                      366
availability_90                        91
bathrooms                              17
bed_type                                5
bedrooms                               12
beds                                   19
calculated_host_listings_count         31
cancellation_policy                     6
cleaning_fee                          134
description                         24140
experiences_offered                     1
extra_people                           65
first_review                         2645
guests_included                        16
has_availability                        1
host_acceptance_rate                   98
host_has_profile_pic                    2
host_identity_verified                  2
host_is_superhost                       2
host_listings_count                    52
host_location                        1095
host_response_rate                

Three main insights from unique values:

- Some columns have only 1 value and can be dropped
- Some other columns have 2 values and appear to be true/false (i.e. can be recoded as 1/0)
- Certain columns have a high number of unique values, which can probably be clustered into a few relevant ones (e.g. cancellation_policy, property_type)

In [23]:
# List missing values (pre-cleaning)


def count_missing(data):
    null_cols = data.columns[data.isnull().any(axis=0)]
    X_null = data[null_cols].isnull().sum()
    X_null = X_null.sort_values(ascending=False)
    print(X_null)


count_missing(data)

square_feet                    24764
monthly_price                  22991
weekly_price                   22138
notes                          17288
interaction                    12577
house_rules                    12575
host_response_rate             12092
host_response_time             12092
neighborhood_overview          10950
transit                         9696
security_deposit                9696
space                           8848
host_acceptance_rate            8508
cleaning_fee                    7439
review_scores_value             5077
review_scores_checkin           5075
review_scores_location          5074
review_scores_communication     5056
review_scores_accuracy          5054
review_scores_cleanliness       5051
review_scores_rating            5027
last_review                     4528
reviews_per_month               4528
first_review                    4528
summary                         1249
zipcode                          513
description                      500
b

Various features have a lot of missing values. In particular, there is an observable cut where many features have more than 4.500 missing values and the rest has less than 1.000. The former - except for review_scores - shall be removed, the latter imputed.

**Conclusions (selection)**

- **host_response_rate** and **host_response_time** are unfortunately not available for half of the dataset and consequently the columns are removed
- **review_scores** are difficult to replace if they do not exist, but at 0 they will distort the modeling. Hence, missing values are set to mean of the column
- listings without **name** and the few rows without enhanced **host information** (e.g. superhost), **bedrooms** or **bathrooms** are removed and not substantial in number
- missing values for **summary** and **description** are replaced with "" and kept in order to calculate length during feature engineering
- several features with missing values will be directly converted to 1/0 for simplification (**house_rules, security_deposit, space, cleaning_fee, monthly_price, weekly_price**)

## Define data cleaning functions

**Handle missing/incorrect values**

In [24]:
# Define function for filling missing/incorrect values

def cln_fill_missing_val(data):
   
    # Convert columns with missing values to 1/0
    data.security_deposit.fillna("0", inplace=True)
    data.cleaning_fee.fillna("0", inplace=True)
    data.monthly_price.fillna("0", inplace=True)
    data.weekly_price.fillna("0", inplace=True)

    
    # Fill missing values of "beds" with 0 and then set all with "bed_type" Real Bed to at least 1, those with value "0" to 0.5
    data.beds.fillna(0, inplace=True)
    data.beds = np.where((data.beds == 0) & (data.bed_type == "Real Bed"), 1,
                         data.beds)
    data.beds = np.where((data.beds == 0), 0.5, data.beds)

    
    # Set all with "bathrooms" and "bedrooms" 0 to at least 0.5
    data.bathrooms = np.where(data.bathrooms == 0, 0.5, data.bathrooms)
    data.bedrooms = np.where(data.bedrooms == 0, 0.5, data.bedrooms)

    
    # Fill review_scores with median
    data.review_scores_rating.fillna(data.review_scores_rating.median(),
                                     inplace=True)
    data.review_scores_value.fillna(data.review_scores_value.median(),
                                    inplace=True)
    data.review_scores_checkin.fillna(data.review_scores_checkin.median(),
                                      inplace=True)
    data.review_scores_location.fillna(data.review_scores_location.median(),
                                       inplace=True)
    data.review_scores_communication.fillna(
        data.review_scores_communication.median(), inplace=True)
    data.review_scores_accuracy.fillna(data.review_scores_accuracy.median(),
                                       inplace=True)
    data.review_scores_cleanliness.fillna(
        data.review_scores_cleanliness.median(), inplace=True)

    
    # Fill host_response/acceptance columns with median/"unknown"
    data.host_response_time.fillna("unknown", inplace=True)
    data.host_acceptance_rate.fillna(data.review_scores_rating.median(),
                                     inplace=True)
    data.host_response_rate.fillna(data.review_scores_rating.median(),
                                   inplace=True)

    
    # Fill missing text values with ""
    data.description.fillna("", inplace=True)
    data.interaction.fillna("", inplace=True)
    data.house_rules.fillna("", inplace=True)
    data.neighborhood_overview.fillna("", inplace=True)
    data.notes.fillna("", inplace=True)
    data.space.fillna("", inplace=True)
    data.summary.fillna("", inplace=True)
    data.transit.fillna("", inplace=True)
    
    return data

**Handle wrong/varying datatypes**

In [25]:
# Define function for changing datatypes

def cln_chg_datatypes(data):
    
    # Convert numeric objects to float
    data.cleaning_fee = [
        float(i.strip("$").replace(",", "")) for i in data.cleaning_fee
    ]
    data.extra_people = [
        float(i.strip("$").replace(",", "")) for i in data.extra_people
    ]
    data.host_acceptance_rate = [
        float(str(i).strip("%")) for i in data.host_acceptance_rate
    ]
    data.host_response_rate = [
        float(str(i).strip("%")) for i in data.host_response_rate
    ]
    data.monthly_price = [
        float(i.strip("$").replace(",", "")) for i in data.monthly_price
    ]
    data.price = [float(i.strip("$").replace(",", "")) for i in data.price]
    data.security_deposit = [
        float(i.strip("$").replace(",", "")) for i in data.security_deposit
    ]
    data.weekly_price = [
        float(i.strip("$").replace(",", "")) for i in data.weekly_price
    ]

    
    # Convert varying zipcode datatypes to string
    data.zipcode = ["zip_" + str(i)[:5] for i in data.zipcode]

    
    # Convert date objects to datetime
    data.first_review = data.first_review.astype('datetime64[D]')
    data.last_review = data.last_review.astype('datetime64[D]')
    
    return data

**Add select amenities as column to data**

In [26]:
# Define function for selecting amenities

def cln_sel_amenities(data):
  
    # Create temporary list with all amenities per listing
    amenities_temp = [
        data.amenities[i].strip("{").strip("}").split(',') for i in data.index
    ]

    
    # Add all amenities to single list in order to count occurrences
    amenities = []
    for lst in amenities_temp:
        for item in lst:
            amenities.append(item)
    amenities = pd.Series(amenities)

    
    # Display count of individual amenities
    #print(amenities.value_counts())

    
    # Add select amenities as distinct columns to data
    data.loc[data.amenities.str.contains('Balcony|Patio'), 'am_balcony'] = 1
    data.am_balcony.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains(
        'Beach view|Beachfront|Lake access|Mountain view|Ski-in/Ski-out|Waterfront'
    ), 'am_nature_and_views'] = 1
    data.am_nature_and_views.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Breakfast'), 'am_breakfast'] = 1
    data.am_breakfast.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('TV'), 'am_tv'] = 1
    data.am_tv.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Coffee maker|Espresso machine'),
             'am_coffee_machine'] = 1
    data.am_coffee_machine.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Cooking basics'),
             'am_cooking_basics'] = 1
    data.am_cooking_basics.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Dishwasher|Dryer|Washer'),
             'am_white_goods'] = 1
    data.am_white_goods.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Elevator'), 'am_elevator'] = 1
    data.am_elevator.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Essentials'), 'am_essentials'] = 1
    data.am_essentials.fillna(0, inplace=True)

    data.loc[
        data.amenities.str.contains('Family/kid friendly|Children|children'),
        'am_child_friendly'] = 1
    data.am_child_friendly.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('parking'), 'am_parking'] = 1
    data.am_parking.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Pets|pet|Cat(s)|Dog(s)'),
             'am_pets_allowed'] = 1
    data.am_pets_allowed.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Private entrance'),
             'am_private_entrance'] = 1
    data.am_private_entrance.fillna(0, inplace=True)

    data.loc[data.amenities.str.contains('Smoking allowed'),
             'am_smoking_allowed'] = 1
    data.am_smoking_allowed.fillna(0, inplace=True)
    
    return data

Out of the full list of amenities, not all will have a significant impact on the price. For the purpose of this analysis, an initial selection has been made and then enhanced by some great [previous work](https://github.com/L-Lewis/Airbnb-neural-network-price-prediction/blob/master/Airbnb-price-prediction.ipynb) on selecting relevant amenities. Additionally, most amenities with a split of more than 90/10 between 1/0 have been **removed (strikethrough in the list)** - except for some that were deemed substantial (24-hour check-in, breakfast, essentials, nature and views)

| **NEW COLUMN** | **PREVIOUS AMENITY/IES** |
| :----- | :----- |
| <s>**am_check_in_24h**</s> | <s>24-hour check-in</s> |
| **<s>am_air_con</s>** | <s>Air conditioning/central air conditioning</s> |
| **am_balcony** | Balcony/patio or balcony |
| **am_nature_and_views** | Beach view/beachfront/lake access/mountain view/ski-in ski-out/waterfront (i.e. great location/views) |
| **am_breakfast** | Breakfast |
| **am_tv** | Cable TV/TV |
| **am_coffee_machine** | Coffee maker/espresso machine |
| **am_cooking_basics** | Cooking basics |
| **am_white_goods** | Dishwasher/Dryer/Washer/Washer and dryer |
| **am_elevator** | Elevator |
| <s>**am_gym**</s> | <s>Exercise equipment/gym/private gym/shared gym</s> |
| **am_essentials** | Essentials |
| **am_child_friendly** | Family/kid friendly, or anything containing 'children' |
| **am_parking** | Free parking on premises/free street parking/outdoor parking/paid parking off premises/paid parking on premises |
| <s>**am_outdoor_space**</s> | <s>Garden or backyard/outdoor seating/sun loungers/terrace</s> |
| <s>**am_wellness**</s> | <s>Hot tub/jetted tub/private hot tub/sauna/shared hot tub/pool/private pool/shared pool</s> |
| <s>**am_internet**</s> | <s>Internet/pocket wifi/wifi</s> |
| **am_pets_allowed** | Pets allowed/cat(s)/dog(s)/pets live on this property/other pet(s) |
| **am_private_entrance** | Private entrance |
| <s>**am_secure**</s> | <s>Safe/security system</s> |
| <s>**am_self_check_in**</s> | <s>Self check-in</s> |
| **am_smoking_allowed** | Smoking allowed |

**Drop irrelevant rows**

Many of the decisions below are based on assumptions and judgement derived from the analysis, EDA and intuition.

In [27]:
# Define function for dropping irrelevant rows

def cln_drop_rows(data):
    
    # Drop missing rows of features with few missing values
    data.dropna(subset=[
        "name", "host_is_superhost", "bedrooms", "bathrooms",
        "neighbourhood_cleansed", "zipcode"
    ],
                inplace=True)

    
    # Drop missing rows of zipcode
    data = data[data.zipcode != "zip_nan"]

    
    # Remove "poor" listings (value above/below a certain threshold)
    data = data[data.price < 500]
    data = data[data.price >= 10]
    data = data[data.minimum_nights <= 100]

    
    # Remove listings where "accommodates" is lower than "guests_included"
    data = data[data.accommodates - data.guests_included >= 0]

    
    # Remove listings where "accommodates" > 10 (outliers)
    data = data[data.accommodates <= 10]

    
    # Remove listings where "accommodates" - "beds" < 0
    data = data[data.accommodates - data.beds >= 0]

    
    # Remove listings where "bedrooms" - "beds" > 2
    data = data[data.bedrooms - data.beds <= 2]

    
    # Remove listings where "beds" - "bedrooms" > 10
    data = data[data.beds - data.bedrooms <= 10]

    
    # Remove listings where "monthly_price" is more than 30x "price"
    data = data[data.monthly_price / data.price <= 30]

    
    # Remove listings where "weekly_price" is more than 7x "price"
    data = data[data.weekly_price / data.price <= 7]

    
    # Remove "inactive" or "new" listings with no reviews in last twelve months
    data = data[data.number_of_reviews_ltm != 0]

    
    # Remove listings with no "availability_365" and no reviews in last three months
    data = data[(data.availability_365 != 0) |
                (data.last_review > dataset_date_min3mth)]
    
    return data

**Drop irrelevant columns**

Remove low-frequency classes from categorical columns: Neighbourhoods_cleansed and zipcodes both have a high volume of unique values, many with under 10 occurrences. Hence, all with <0.25% share are bundled as "other".

In [28]:
# Define function for dropping irrelevant columns

def cln_drop_cols(data):    
    
    # Change neighbourhoods_cleansed that make up <0.25% of data to "other"
    data = data.apply(lambda x: x.mask(
        x.map(x.value_counts()) < (0.0025 * len(data)), 'nb_other')
                      if x.name == 'neighbourhood_cleansed' else x)
    
    
    # Change zipcodes that make up <0.25% of data to "other"
    data = data.apply(lambda x: x.mask(
        x.map(x.value_counts()) < (0.0025 * len(data)), 'zip_other')
                      if x.name == 'zipcode' else x)
    
    
    # Drop irrelevant columns
    data.drop(
        [
            "bed_type",
            "experiences_offered",
            "has_availability",
    #        "host_acceptance_rate",
            "host_location",
    #        "host_response_rate",
    #        "host_response_time",  
    #        "number_of_reviews", 
    #        "number_of_reviews_ltm",
            "requires_license",
            "is_business_travel_ready",
            "host_has_profile_pic",
            "host_listings_count",
            "require_guest_profile_picture",
            "require_guest_phone_verification",
            "reviews_per_month",
            "square_feet"
        ],
        inplace=True,
        axis=1)
    
    return data

Explanation for selection of dropped columns:

| **FEATURE(S)** | **NOTES** |
| :----- | :----- | 
| **bed_type** | over 97% of values were "Real Bed", hence little added value |
| **experiences_offered** | all values are "none" |
| **has_availability** | all values are "t" |
| **requires_license, host_has_profile_pic** | almost all values are "t" |
| **is_business_travel_ready** | all values are "f" |
| **require_guest_xxx** | almost all values are "f" |
| **host_listings_count** | calculated_host_listings_count appears to be a sanitized version (ranges from 1 to 55) of host_listings_count (has values 0 and highest is 1397) |
 **other host_xyz** | too many missing values |
| **reviews_per_month** | number_of_reviews_ltm kept as better measure |
| **square_feet** | too many missing values |
| <s>**property_type**</s> | <s>90% of values are "apartment", too many unique values to sensibly classify</s> kept instead |

## Apply data cleaning functions

In [29]:
# Bundle data cleaning steps as function "data_cleaning"
def data_cleaning(data):
    data = cln_fill_missing_val(data)
    data = cln_chg_datatypes(data)
    data = cln_sel_amenities(data)
    data = cln_drop_rows(data)
    data = cln_drop_cols(data)
    
    return data

In [30]:
# Apply data cleaning to dataset
data = data_cleaning(data)

C:\Users\Mauricio\anaconda3\envs\airbnb42\lib\site-packages\pandas\core\strings.py:2001: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


## Final Check, Cleaning and Export

In [31]:
# Sort columns in dataset
data = data.reindex(sorted(data.columns, reverse=False), axis=1)

In [32]:
# List datatypes (data.info()) (post-cleaning)
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10471 entries, 3176 to 42885615
Data columns (total 66 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   accommodates                    10471 non-null  int64         
 1   am_balcony                      10471 non-null  float64       
 2   am_breakfast                    10471 non-null  float64       
 3   am_child_friendly               10471 non-null  float64       
 4   am_coffee_machine               10471 non-null  float64       
 5   am_cooking_basics               10471 non-null  float64       
 6   am_elevator                     10471 non-null  float64       
 7   am_essentials                   10471 non-null  float64       
 8   am_nature_and_views             10471 non-null  float64       
 9   am_parking                      10471 non-null  float64       
 10  am_pets_allowed                 10471 non-null  float64       
 

Many columns are only kept for EDA or feature engineering and will be dropped afterwards. The remaining columns are prepared and ready for further processing.

In [33]:
# List missing values (post-cleaning)

def count_missing(data):
    null_cols = data.columns[data.isnull().any(axis=0)]
    X_null = data[null_cols].isnull().sum()
    X_null = X_null.sort_values(ascending=False)
    print(X_null)

count_missing(data)
#data.isnull().sum()

Series([], dtype: float64)


As we can see, we got rid of all the missing values

In [34]:
# Display cleaned dataset
print(data.shape)
data.head(3)

(10471, 66)


accommodates  am_balcony  am_breakfast  am_child_friendly  \
id                                                                
3176             4         0.0           0.0                1.0   
3309             1         0.0           0.0                0.0   
6883             2         0.0           0.0                0.0   

      am_coffee_machine  am_cooking_basics  am_elevator  am_essentials  \
id                                                                       
3176                0.0                0.0          0.0            1.0   
3309                0.0                0.0          0.0            1.0   
6883                1.0                1.0          1.0            1.0   

      am_nature_and_views  am_parking  am_pets_allowed  am_private_entrance  \
id                                                                            
3176                  0.0         0.0              0.0                  0.0   
3309                  0.0         1.0              1.0                  0.0   
6883                  0.0         0.0              0.0                  0.0   

      am_smoking_allowed  am_tv  am_white_goods  \
id                                                
3176                 0.0    0.0             1.0   
3309                 0.0    0.0             1.0   
6883                 0.0    1.0             1.0   

                                              amenities  availability_365  \
id                                                                          
3176  {Internet,Wifi,Kitchen,"Buzzer/wireless interc...               140   
3309  {Internet,Wifi,"Pets live on this property",Ca...               320   
6883  {TV,"Cable TV",Internet,Wifi,"Air conditioning...                 0   

      availability_90  bathrooms  bedrooms  beds  \
id                                                 
3176                0        1.0       1.0   2.0   
3309               45        1.0       1.0   1.0   
6883                0        1.0       1.0   1.0   

      calculated_host_listings_count          cancellation_policy  \
id                                                                  
3176                               1  strict_14_with_grace_period   
3309                               1  strict_14_with_grace_period   
6883                               1                     moderate   

      cleaning_fee                                        description  \
id                                                                      
3176         100.0  This beautiful first floor apartment  is situa...   
3309          30.0  First of all: I prefer short-notice bookings. ...   
6883          39.0  Stay in a stylish loft on the second floor and...   

      extra_people first_review  guests_included  host_acceptance_rate  \
id                                                                       
3176          20.0   2009-06-20                2                  80.0   
3309          18.0   2013-08-12                1                 100.0   
6883           0.0   2010-02-15                1                 100.0   

     host_identity_verified host_is_superhost  host_response_rate  \
id                                                                  
3176                      t                 f               100.0   
3309                      f                 f               100.0   
6883                      t                 f               100.0   

      host_response_time                                        house_rules  \
id                                                                            
3176  within a few hours  It’s a non smoking flat, which likes to be tre...   
3309        within a day  House-Rules and Information ..............(deu...   
6883      within an hour  No Pets.  No loud Parties.  Smoking only on th...   

     instant_bookable                                        interaction  \
id                                                                         
3176                f  

The number of listings has been reduced substantially by removing rows deemed to be irrelevant or distorting.

**Export data_clean**

In [35]:
# Create path to export dataset (if not existing)
if not os.path.exists(f"saves/{dataset_loc}_{dataset_date}/"):
    os.mkdir(f"saves/{dataset_loc}_{dataset_date}/")

In [36]:
# Export dataset for further use in 2_EDA_Clean and 3_Feature_Engineering
data_clean = data.copy()
data_clean.to_pickle(f"saves/{dataset_loc}_{dataset_date}/data_clean.pkl")